ibis/pandas/core.py

```diff
@@ -554,3 +610,113 @@ def canonicalize_context(
             f'begin time {begin} must be before or equal' f' to end time {end}'
         )
     return begin, end
+
+
+class TimeContextRelation(enum.Enum):
+    """ Enum to classify the relationship between two time contexts
+        Assume that we have two timecontext c1 (begin1, end1),
+        c2(begin2, end2):
+        SUBSET means c1 is a subset of c2, begin1 is greater than or equal to
+        begin2, and end1 is less than or equal to end2.
+        Likewise, SUPERSET means that begin1 is earlier than begin2, and end1
+        is later than end2.
+        If neither of the two contexts is a superset of each other, and they
+        share some time range in common, we called them OVERLAP.
+        And NONOVERLAP means the two contexts doesn't overlap at all, which
+        means end1 is earlier than begin2 or end2 is earlier than begin1
+    """
+
+    SUBSET = 0
+    SUPERSET = 1
+    OVERLAP = 2
+    NONOVERLAP = 3
+
+
+def compare_timecontext(cur_context: TimeContext, old_context: TimeContext):
+    """Compare two timecontext and return the relationship between two time
+    context (SUBSET, SUPERSET, OVERLAP, NONOVERLAP).
+
+    Parameters
+    ----------
+    cur_context: TimeContext
+    old_context: TimeContext
+
+    Returns
+    -------
+    result : Enum[TimeContextRelation]
+    """
+    begin, end = cur_context
+    old_begin, old_end = old_context
+    if old_begin <= begin and old_end >= end:
+        return TimeContextRelation.SUBSET
+    elif old_begin >= begin and old_end <= end:
+        return TimeContextRelation.SUPERSET
+    elif old_end < begin or end < old_begin:
+        return TimeContextRelation.NONOVERLAP
+    else:
+        return TimeContextRelation.OVERLAP
+
+
+def scope_item(op, result, timecontext):
```

Comment:
Can we lift scope related code to its own module so the pyspark backend can also reuse it?

Reply:
Same as above, will move this to Ibis level